contracts/buy/PartyBuy.sol

```diff
@@ -0,0 +1,180 @@
+
+    // emitted if the Party fails to buy the token before expiresAt
+    // and someone closes the Party so folks can reclaim ETH
+    event Closed();
+
+    // ======== Constructor =========
+
+    constructor(
+        address _partyDAOMultisig,
+        address _tokenVaultFactory,
+        address _weth
+    ) Party(_partyDAOMultisig, _tokenVaultFactory, _weth) {}
+
+    // ======== Initializer =========
+
+    function initialize(
+        address _nftContract,
+        uint256 _tokenId,
+        uint256 _maxPrice,
+        uint256 _secondsToTimeout,
+        address _splitRecipient,
+        uint256 _splitBasisPoints,
+        string memory _name,
+        string memory _symbol
+    ) external initializer {
+        // initialize ReentrancyGuard and ERC721Holder
+        __ReentrancyGuard_init();
+        __ERC721Holder_init();
+        // set storage variables
+        nftContract = IERC721Metadata(_nftContract);
+        tokenId = _tokenId;
+        name = _name;
+        symbol = _symbol;
+        expiresAt = _secondsToTimeout + block.timestamp;
+        require(_maxPrice > 0, "PartyBuy::initialize: must set price higher than 0");
+        maxPrice = _maxPrice;
+        // validate that party split won't retain the total token supply
+        uint256 _remainingBasisPoints = 10000 - TOKEN_FEE_BASIS_POINTS;
+        require(_splitBasisPoints < _remainingBasisPoints, "PartyBuy::initialize: basis points can't take 100%");
+        // validate that a portion of the token supply is not being burned
+        if (_splitRecipient == address(0)) {
+            require(_splitBasisPoints == 0, "PartyBuy::initialize: can't send tokens to burn addr");
+        }
+        splitBasisPoints = _splitBasisPoints;
+        splitRecipient = _splitRecipient;
+        // validate token exists
+        require(_getOwner() != address(0), "PartyBuy::initialize: NFT getOwner failed");
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
```

Comment:
can we introduce an allow-list so only known good targetContracts can be used?

Reply:
we could, but this is more of a product question (cc @johncpalmer )

Reply:
i would feel significantly more safe if we did this

Reply:
👍   let's do it